src/compiler/checker.ts

```diff
@@ -33082,28 +33307,13 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         else {
             if (isAnyLike) {
                 if (isIdentifier(left) && parentSymbol) {
-                    markAliasReferenced(parentSymbol, node);
+                    markLinkedReferences(node);
                 }
                 return isErrorType(apparentType) ? errorType : apparentType;
             }
             prop = getPropertyOfType(apparentType, right.escapedText, /*skipObjectFunctionPropertyAugment*/ isConstEnumObjectType(apparentType), /*includeTypeOnlyMembers*/ node.kind === SyntaxKind.QualifiedName);
         }
-        // In `Foo.Bar.Baz`, 'Foo' is not referenced if 'Bar' is a const enum or a module containing only const enums.
-        // `Foo` is also not referenced in `enum FooCopy { Bar = Foo.Bar }`, because the enum member value gets inlined
-        // here even if `Foo` is not a const enum.
-        //
-        // The exceptions are:
-        //   1. if 'isolatedModules' is enabled, because the const enum value will not be inlined, and
-        //   2. if 'preserveConstEnums' is enabled and the expression is itself an export, e.g. `export = Foo.Bar.Baz`.
-        if (
-            isIdentifier(left) && parentSymbol && (
-                getIsolatedModules(compilerOptions) ||
-                !(prop && (isConstEnumOrConstEnumOnlyModule(prop) || prop.flags & SymbolFlags.EnumMember && node.parent.kind === SyntaxKind.EnumMember)) ||
-                shouldPreserveConstEnums(compilerOptions) && isExportOrExportExpression(node)
-            )
-        ) {
-            markAliasReferenced(parentSymbol, node);
-        }
+        markLinkedReferences(node, prop, leftType);

```

Comment:
I'm hoping reusing the `leftType`+`prop` here negates the performance impact this change had. Non-checker uses aren't expected to have done this work already, so it does still need to have the logic to lookup the `prop` itself, and is probably the costly part of alias marking. That you can't just say `A.B;` references `A` because `A.B` might be a `const enum` reference that gets inlined (and so you need to typecheck) is actually quite annoying from an encapsulation point of view - it's the _one place_ in this process where we have to resort to typechecking. I guess that's why it's not done in `isolatedModules`. 😄